File: ring-android/app/src/main/java/cx/ring/fragments/GeneralAccountFragment.java
/*
 *  Copyright (C) 2004-2021 Savoir-faire Linux Inc.
 *
 *  Author: Alexandre Lision <alexandre.lision@savoirfairelinux.com>
 *          Adrien Béraud <adrien.beraud@savoirfairelinux.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package cx.ring.fragments;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.text.format.Formatter;
import android.view.inputmethod.EditorInfo;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.FragmentManager;
import androidx.preference.EditTextPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceManager;
import androidx.preference.SeekBarPreference;
import androidx.preference.SwitchPreference;
import androidx.preference.TwoStatePreference;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import cx.ring.R;
import cx.ring.account.AccountEditionFragment;
import cx.ring.application.JamiApplication;
import net.jami.model.Account;
import net.jami.model.AccountConfig;
import net.jami.model.ConfigKey;
import cx.ring.mvp.BasePreferenceFragment;
import cx.ring.services.SharedPreferencesServiceImpl;
import net.jami.utils.Log;
import net.jami.utils.Tuple;
import cx.ring.views.EditTextIntegerPreference;
import cx.ring.views.EditTextPreferenceDialog;
import cx.ring.views.PasswordPreference;
import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class GeneralAccountFragment extends BasePreferenceFragment<GeneralAccountPresenter> implements GeneralAccountView {

    public static final String TAG = GeneralAccountFragment.class.getSimpleName();

    private static final String DIALOG_FRAGMENT_TAG = "androidx.preference.PreferenceFragment.DIALOG";
    private final Preference.OnPreferenceChangeListener changeAccountStatusListener = (preference, newValue) -> {
        presenter.setEnabled((Boolean) newValue);
        return false;
    };
    private final Preference.OnPreferenceChangeListener changeBasicPreferenceListener = (preference, newValue) -> {
        Log.i(TAG, "Changing preference " + preference.getKey() + " to value:" + newValue);
        final ConfigKey key = ConfigKey.fromString(preference.getKey());
        if (preference instanceof TwoStatePreference) {
            presenter.twoStatePreferenceChanged(key, newValue);
        } else if (preference instanceof PasswordPreference) {
            StringBuilder tmp = new StringBuilder();
            for (int i = 0; i < ((String) newValue).length(); ++i) {
                tmp.append("*");
            }
            preference.setSummary(tmp.toString());
            presenter.passwordPreferenceChanged(key, newValue);
        } else if (key == ConfigKey.ACCOUNT_USERNAME) {
            presenter.userNameChanged(key, newValue);
            preference.setSummary((CharSequence) newValue);
        } else {
            preference.setSummary((CharSequence) newValue);
            presenter.preferenceChanged(key, newValue);
        }
        return true;
    };

    public static GeneralAccountFragment newInstance(@NonNull String accountId) {
        Bundle bundle = new Bundle();
        bundle.putString(AccountEditionFragment.ACCOUNT_ID_KEY, accountId);
        GeneralAccountFragment generalAccountFragment = new GeneralAccountFragment();
        generalAccountFragment.setArguments(bundle);
        return generalAccountFragment;
    }

    @Override
    public void accountChanged(@NonNull Account account) {
        PreferenceManager pm = getPreferenceManager();
        pm.setSharedPreferencesMode(Context.MODE_PRIVATE);
        pm.setSharedPreferencesName(SharedPreferencesServiceImpl.PREFS_ACCOUNT+account.getAccountID());

        setPreferenceDetails(account.getConfig());

        SwitchPreference pref = findPreference("Account.status");
        if (account.isSip() && pref != null) {
            String status;
            pref.setTitle(account.getAlias());
            if (account.isEnabled()) {
                if (account.isTrying()) {
                    status = getString(R.string.account_status_connecting);
                } else if (account.needsMigration()) {
                    status = getString(R.string.account_update_needed);
                } else if (account.isInError()) {
                    status = getString(R.string.account_status_connection_error);
                } else if (account.isRegistered()) {
                    status = getString(R.string.account_status_online);
                } else {
                    status = getString(R.string.account_status_unknown);
                }
            } else {
                status = getString(R.string.account_status_offline);
            }
            pref.setSummary(status);
            pref.setChecked(account.isEnabled());

            // An ip2ip account is always ready
            pref.setEnabled(!account.isIP2IP());

            pref.setOnPreferenceChangeListener(changeAccountStatusListener);
        }

        setPreferenceListener(account.getConfig(), changeBasicPreferenceListener);
    }

    @Override
    public void finish() {
        Activity activity = getActivity();
        if (activity != null)
            activity.onBackPressed();
    }

    @Override
    public void updateResolutions(Tuple<Integer, Integer> maxResolution, int currentResolution) {
    }

    private CharSequence getFileSizeSummary(int size, int maxSize) {
        if (size == 0)  {
            return getText(R.string.account_accept_files_never);
        } else if (size == maxSize) {
            return getText(R.string.account_accept_files_always);
        } else {
            return Formatter.formatFileSize(requireContext(), size * 1000 * 1000);
        }
    }

    @Override
    public void onCreatePreferences(Bundle bundle, String rootKey) {
        super.onCreatePreferences(bundle, rootKey);

        Bundle args = getArguments();
        presenter.init(args == null  ? null : args.getString(AccountEditionFragment.ACCOUNT_ID_KEY));

        SeekBarPreference filePref = findPreference("acceptIncomingFilesMaxSize");
        if (filePref != null) {
            filePref.setOnPreferenceChangeListener((p, v) ->  {
                SeekBarPreference pref = (SeekBarPreference)p;
                p.setSummary(getFileSizeSummary((Integer) v, pref.getMax()));
                return true;
            });
            filePref.setSummary(getFileSizeSummary(filePref.getValue(), filePref.getMax()));
        }

        Preference deletePref = findPreference("Account.delete");
        if (deletePref != null) {
            deletePref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    AlertDialog deleteDialog = createDeleteDialog();
                    deleteDialog.show();
                    return false;
                }
            });
        }
    }

    @Override
    public void onDisplayPreferenceDialog(Preference preference) {
        FragmentManager fragmentManager = getParentFragmentManager();
        if (fragmentManager.findFragmentByTag(DIALOG_FRAGMENT_TAG) != null) {
            return;
        }

        if (preference instanceof EditTextIntegerPreference) {
            EditTextPreferenceDialog f = EditTextPreferenceDialog.newInstance(preference.getKey(), EditorInfo.TYPE_CLASS_NUMBER);
            f.setTargetFragment(this, 0);
            f.show(fragmentManager, DIALOG_FRAGMENT_TAG);
        } else if (preference instanceof PasswordPreference) {
            EditTextPreferenceDialog f = EditTextPreferenceDialog.newInstance(preference.getKey(), EditorInfo.TYPE_CLASS_TEXT | EditorInfo.TYPE_TEXT_VARIATION_PASSWORD);
            f.setTargetFragment(this, 0);
            f.show(fragmentManager, DIALOG_FRAGMENT_TAG);
        } else {
            super.onDisplayPreferenceDialog(preference);
        }
    }

    private void setPreferenceDetails(AccountConfig details) {
        for (ConfigKey confKey : details.getKeys()) {
            Preference pref = findPreference(confKey.key());
            if (pref == null) {
                continue;
            }
            if (!confKey.isTwoState()) {
                String val = details.get(confKey);
                ((EditTextPreference) pref).setText(val);
                if (pref instanceof PasswordPreference) {
                    StringBuilder tmp = new StringBuilder();
                    for (int i = 0; i < val.length(); ++i) {
                        tmp.append("*");
                    }
                    pref.setSummary(tmp.toString());
                } else {
                    pref.setSummary(val);
                }
            } else {
                ((TwoStatePreference) pref).setChecked(details.getBool(confKey));
            }
        }
    }

    private void setPreferenceListener(AccountConfig details, Preference.OnPreferenceChangeListener listener) {
        for (ConfigKey confKey : details.getKeys()) {
            Preference pref = findPreference(confKey.key());
            if (pref != null) {
                pref.setOnPreferenceChangeListener(listener);
            }
        }
    }

    @Override
    public void addJamiPreferences(String accountId) {
        PreferenceManager pm = getPreferenceManager();
        pm.setSharedPreferencesMode(Context.MODE_PRIVATE);
        pm.setSharedPreferencesName(SharedPreferencesServiceImpl.PREFS_ACCOUNT+accountId);
        addPreferencesFromResource(R.xml.account_prefs_jami);
    }

    @Override
    public void addSipPreferences() {
        addPreferencesFromResource(R.xml.account_general_prefs);
    }

    @NonNull
    private AlertDialog createDeleteDialog() {
        AlertDialog alertDialog = new MaterialAlertDialogBuilder(requireContext())
                .setMessage(R.string.account_delete_dialog_message)
                .setTitle(R.string.account_delete_dialog_title)
                .setPositiveButton(R.string.menu_delete, (dialog, whichButton) -> presenter.removeAccount())
                .setNegativeButton(android.R.string.cancel, null)
                .create();
        Activity activity = getActivity();
        if (activity != null)
            alertDialog.setOwnerActivity(getActivity());
        return alertDialog;
    }

}

File: jami-android/app/src/main/java/cx/ring/fragments/GeneralAccountFragment.kt
/*
 *  Copyright (C) 2004-2024 Savoir-faire Linux Inc.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package cx.ring.fragments

import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.text.format.Formatter
import android.util.Log
import android.view.View
import android.view.inputmethod.EditorInfo
import androidx.appcompat.app.AlertDialog
import androidx.preference.*
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import cx.ring.R
import cx.ring.account.AccountEditionFragment
import cx.ring.interfaces.AppBarStateListener
import cx.ring.mvp.BasePreferenceFragment
import cx.ring.services.SharedPreferencesServiceImpl
import cx.ring.views.EditTextIntegerPreference
import cx.ring.views.EditTextPreferenceDialog
import cx.ring.views.PasswordPreference
import dagger.hilt.android.AndroidEntryPoint
import net.jami.model.Account
import net.jami.model.AccountConfig
import net.jami.model.ConfigKey
import net.jami.model.ConfigKey.Companion.fromString
import net.jami.settings.GeneralAccountPresenter
import net.jami.settings.GeneralAccountView

@AndroidEntryPoint
class GeneralAccountFragment : BasePreferenceFragment<GeneralAccountPresenter>(), GeneralAccountView {
    private val changeAccountStatusListener = Preference.OnPreferenceChangeListener { preference: Preference?, newValue: Any ->
        presenter.setEnabled(newValue as Boolean)
        false
    }
    private val changeBasicPreferenceListener = Preference.OnPreferenceChangeListener { preference: Preference, newValue: Any ->
        Log.i(TAG, "Changing preference ${preference.key} type ${preference.javaClass.simpleName} to value:$newValue")
        val key = fromString(preference.key) ?: return@OnPreferenceChangeListener false
        if (preference is TwoStatePreference) {
            presenter.twoStatePreferenceChanged(key, newValue)
        } else if (preference is PasswordPreference) {
            preference.setSummary((newValue as CharSequence).map { "•" }.joinToString(""))
            presenter.passwordPreferenceChanged(key, newValue)
        } else if (key === ConfigKey.ACCOUNT_USERNAME) {
            presenter.userNameChanged(key, newValue)
            preference.summary = newValue as CharSequence
        } else {
            preference.summary = newValue as CharSequence
            presenter.preferenceChanged(key, newValue)
        }
        true
    }

    override fun accountChanged(account: Account) {
        val pm = preferenceManager
        pm.sharedPreferencesMode = Context.MODE_PRIVATE
        pm.sharedPreferencesName = SharedPreferencesServiceImpl.PREFS_ACCOUNT + account.accountId
        setPreferenceDetails(account.config)
        val pref = findPreference<SwitchPreference>("Account.status")
        if (account.isSip && pref != null) {
            pref.title = account.alias
            val status: String = getString(if (account.isEnabled) {
                when {
                    account.isTrying -> R.string.account_status_connecting
                    account.needsMigration() -> R.string.account_update_needed
                    account.isInError -> R.string.account_status_connection_error
                    account.isRegistered -> R.string.account_status_online
                    else -> R.string.account_status_unknown
                }
            } else {
                R.string.account_status_offline
            })
            pref.summary = status
            pref.isChecked = account.isEnabled

            // An ip2ip account is always ready
            pref.isEnabled = !account.isIP2IP
            pref.onPreferenceChangeListener = changeAccountStatusListener
        }
        setPreferenceListener(account.config, changeBasicPreferenceListener)
    }

    override fun finish() {
        activity?.onBackPressedDispatcher?.onBackPressed()
    }

    override fun updateResolutions(maxResolution: Pair<Int, Int>?, currentResolution: Int) {}
    private fun getFileSizeSummary(size: Int, maxSize: Int): CharSequence {
        return if (size == 0) {
            getText(R.string.account_accept_files_never)
        } else if (size == maxSize) {
            getText(R.string.account_accept_files_always)
        } else {
            Formatter.formatFileSize(requireContext(), size * 1000L * 1000L)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        (parentFragment as? AppBarStateListener)?.onAppBarScrollTargetViewChanged(listView)
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        super.onCreatePreferences(savedInstanceState, rootKey)
        val args = arguments
        presenter.init(args?.getString(AccountEditionFragment.ACCOUNT_ID_KEY))
        val filePref = findPreference<SeekBarPreference>("acceptIncomingFilesMaxSize")
        if (filePref != null) {
            filePref.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { p: Preference, v: Any ->
                val pref = p as SeekBarPreference
                p.setSummary(getFileSizeSummary(v as Int, pref.max))
                true
            }
            filePref.summary = getFileSizeSummary(filePref.value, filePref.max)
        }
        val deletePref = findPreference<Preference>("Account.delete")
        if (deletePref != null) {
            deletePref.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                val deleteDialog = createDeleteDialog()
                deleteDialog.show()
                false
            }
        }
    }

    override fun onDisplayPreferenceDialog(preference: Preference) {
        val fragmentManager = parentFragmentManager
        if (fragmentManager.findFragmentByTag(DIALOG_FRAGMENT_TAG) != null) {
            return
        }
        if (preference is EditTextIntegerPreference) {
            val f = EditTextPreferenceDialog.newInstance(preference.getKey(), EditorInfo.TYPE_CLASS_NUMBER)
            f.setTargetFragment(this, 0)
            f.show(fragmentManager, DIALOG_FRAGMENT_TAG)
        } else if (preference is PasswordPreference) {
            val f = EditTextPreferenceDialog.newInstance(
                preference.getKey(),
                EditorInfo.TYPE_CLASS_TEXT or EditorInfo.TYPE_TEXT_VARIATION_PASSWORD
            )
            f.setTargetFragment(this, 0)
            f.show(fragmentManager, DIALOG_FRAGMENT_TAG)
        } else {
            super.onDisplayPreferenceDialog(preference)
        }
    }

    private fun setPreferenceDetails(details: AccountConfig) {
        for (confKey in details.keys) {
            val pref = findPreference<Preference>(confKey.key) ?: continue
            if (!confKey.isBool) {
                val value = details[confKey]
                (pref as EditTextPreference).text = value
                if (pref is PasswordPreference) {
                    pref.setSummary(value.map { "•" }.joinToString(""))
                } else {
                    pref.setSummary(value)
                }
            } else {
                (pref as TwoStatePreference).isChecked = details.getBool(confKey)
            }
        }
    }

    private fun setPreferenceListener(details: AccountConfig, listener: Preference.OnPreferenceChangeListener) {
        for (confKey in details.keys) {
            val pref = findPreference<Preference>(confKey.key)
            if (pref != null) {
                pref.onPreferenceChangeListener = listener
            }
        }
    }

    override fun addJamiPreferences(accountId: String) {
        val pm = preferenceManager
        pm.sharedPreferencesMode = Context.MODE_PRIVATE
        pm.sharedPreferencesName = SharedPreferencesServiceImpl.PREFS_ACCOUNT + accountId
        addPreferencesFromResource(R.xml.account_prefs_jami)
    }

    override fun addSipPreferences() {
        addPreferencesFromResource(R.xml.account_general_prefs)
    }

    private fun createDeleteDialog(): AlertDialog {
        val alertDialog = MaterialAlertDialogBuilder(requireContext())
            .setMessage(R.string.account_delete_dialog_message)
            .setTitle(R.string.account_delete_dialog_title)
            .setPositiveButton(R.string.menu_delete) { dialog: DialogInterface?, whichButton: Int -> presenter.removeAccount() }
            .setNegativeButton(android.R.string.cancel, null)
            .create()
        val activity: Activity? = activity
        if (activity != null) alertDialog.setOwnerActivity(activity)
        return alertDialog
    }

    companion object {
        val TAG = GeneralAccountFragment::class.simpleName!!
        private const val DIALOG_FRAGMENT_TAG = "androidx.preference.PreferenceFragment.DIALOG"
        fun newInstance(accountId: String) = GeneralAccountFragment().apply {
            arguments = Bundle().apply { putString(AccountEditionFragment.ACCOUNT_ID_KEY, accountId) }
        }
    }
}